### PR TITLE
added weightnorm layer

### LINKED ISSUE #3400: Feature request: add weight normalization layer wrapper to Flax
Internal request to add a weight normalization wrapper to Flax.
Can be found in equation 2 in https://arxiv.org/pdf/1602.07868.pdf.

### PR DESCRIPTION
**Pull request by chiamp**
Resolves #3400.

Implemented `nn.WeightNorm` as a layer wrapper. Check out the API documentation [here](https://flax--3405.org.readthedocs.build/en/3405/api_reference/flax.linen/layers.html#flax.linen.WeightNorm).

Example usage:
```
class Baz(nn.Module):
  @nn.compact
  def __call__(self, x):
    return nn.Dense(2)(x)

class Bar(nn.Module):
  @nn.compact
  def __call__(self, x):
    x = Baz()(x)
    x = nn.Dense(3)(x)
    x = Baz()(x)
    x = nn.Dense(3)(x)
    return x

class Foo(nn.Module):
  @nn.compact
  def __call__(self, x):
    x = nn.Dense(3)(x)
    # l2-normalize all params of the second Dense layer
    x = nn.WeightNorm(nn.Dense(4), variable_filter=None)(x)
    x = nn.Dense(5)(x)
    # l2-normalize all kernels in the Bar submodule and all params in the Baz submodule
    x = nn.WeightNorm(Bar(), variable_filter={'kernel', 'Baz'})(x)
    return x

# init
x = jnp.ones((1, 2))
y = jnp.ones((1, 5))
model = Foo()
variables = model.init(jax.random.key(0), x)
variables
```
```
{
  params: {
    ...
    WeightNorm_0: {
        Dense_1/bias/scale: Array([1., 1., 1., 1.], dtype=float32),
        Dense_1/kernel/scale: Array([1., 1., 1., 1.], dtype=float32),
    },
    ...
    WeightNorm_1: {
        Bar_0/Baz_0/Dense_0/bias/scale: Array([1., 1.], dtype=float32),
        Bar_0/Baz_0/Dense_0/kernel/scale: Array([1., 1.], dtype=float32),
        Bar_0/Baz_1/Dense_0/bias/scale: Array([1., 1.], dtype=float32),
        Bar_0/Baz_1/Dense_0/kernel/scale: Array([1., 1.], dtype=float32),
        Bar_0/Dense_0/kernel/scale: Array([1., 1., 1.], dtype=float32),
        Bar_0/Dense_1/kernel/scale: Array([1., 1., 1.], dtype=float32),
    },
    ...
  }
}
```

### CODE REVIEW

File: flax/linen/normalization.py [hunk 1]
```diff
@@ -849,3 +849,169 @@ def _spectral_normalize(self, path, vs, layer_instance_name, update_stats):
...
+    value. This is intended to be a helper function used in this Module's
+    ``__call__`` method in conjunction with ``nn.transforms.map_variables``
+    and ``jax.tree_util.tree_map_with_path``.
+
+    Args:
+      path: dict key path, used for naming the ``scale`` variable
+      vs: variables to be l2-normalized
+    """
+    value = jnp.asarray(vs)
+    str_path = self.layer_instance.name + '/' + '/'.join((dict_key.key for dict_key in path[1:])) + '/scale'
```
Comment (cgarciae) [inline flax/linen/normalization.py:983]: By adding `scale` at the end of the path here, all variables will be normalized if the user is trying to normalize the `scale` variable.

File: flax/linen/normalization.py [hunk 2]
```diff
@@ -849,3 +849,169 @@ def _spectral_normalize(self, path, vs, layer_instance_name, update_stats):
...
+
+    feature_shape = [1] * value.ndim
+    reduced_feature_shape = []
+    for ax in feature_axes:
+      feature_shape[ax] = value.shape[ax]
+      reduced_feature_shape.append(value.shape[ax])
+
+    value_bar = jnp.where(value != 0,
+                          _l2_normalize(value, axis=reduction_axes, eps=self.epsilon),
+                          value)
```
Comment (cgarciae) [inline flax/linen/normalization.py:1006]: Can you explain what is going on here? Seems to be keeping the zeros but now the norm is not `1`.

File: tests/linen/linen_test.py [hunk 3]
```diff
@@ -455,6 +455,148 @@ def __call__(self, x, train):
       else:
         variables = model_cls.init(random.PRNGKey(0), x, train=False)
 
+  def test_manual_weight_norm(self):
+    for feature_axes, reduction_axes in ((-1, 0), (0, 1), ((0, 1), ()), ((), (0, 1)), (None, (0, 1))):
```
Comment (cgarciae) [inline tests/linen/linen_test.py:459]: maybe we can use a parametrized test instead of doing a for loop here?

File: tests/linen/linen_test.py [hunk 4]
```diff
@@ -455,6 +455,148 @@ def __call__(self, x, train):
...
+        return nn.Dense(2)(x)
+    class Bar(nn.Module):
+      @nn.compact
+      def __call__(self, x):
+        x = Baz()(x)
+        x = nn.Dense(3)(x)
+        x = Baz()(x)
+        x = nn.Dense(3)(x)
+        return x
+    for variable_filters, key_paths in (
```
Comment (cgarciae) [inline tests/linen/linen_test.py:490]: also here?

File: flax/linen/normalization.py (hunk 1 above)
Comment (chiamp) [reply flax/linen/normalization.py:983]: Should I keep the keypath and just remove `scale` from the end of it, or should I rename it to something like `weight_scale`?

File: flax/linen/normalization.py (hunk 2 above)
Comment (chiamp) [reply flax/linen/normalization.py]: I removed it

File: flax/linen/normalization.py [hunk 5]
```diff
@@ -849,3 +849,167 @@ def _spectral_normalize(self, path, vs, layer_instance_name, update_stats):
...
+    for ax in feature_axes:
+      feature_shape[ax] = value.shape[ax]
+      reduced_feature_shape.append(value.shape[ax])
+
+    value_bar = _l2_normalize(value, axis=reduction_axes, eps=self.epsilon)
+
+    args = [vs]
+    if self.use_scale:
+      scale = self.param(
+        str_path, self.scale_init, reduced_feature_shape, self.param_dtype
```
Comment (cgarciae) [inline flax/linen/normalization.py:1009]: @chiamp maybe add `/scale` here?

**Review by cgarciae — APPROVED** [review]
